### ISSUE TITLE
Background jobs broken in master under WSL

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#4778]
I'm not sure when this happened, but:

```fish
mqudsi@ZBook ~/r/fish-shell> echo $version
2.7.1-864-g71329a2
mqudsi@ZBook ~/r/fish-shell> cat &
mqudsi@ZBook ~/r/fish-shell>
Job 1, 'cat &' has ended
mqudsi@ZBook ~/r/fish-shell> cat &
mqudsi@ZBook ~/r/fish-shell>
Job 1, 'cat &' has ended
mqudsi@ZBook ~/r/fish-shell>
```

versus in fish 2.7.0:

```fish
mqudsi@ZBook ~/r/fish-shell> fish2/fish
mqudsi@ZBook ~/r/fish-shell> cat &
mqudsi@ZBook ~/r/fish-shell>
Job 1, 'cat &' has stopped
mqudsi@ZBook ~/r/fish-shell> cat &
mqudsi@ZBook ~/r/fish-shell>
Job 2, 'cat &' has stopped
mqudsi@ZBook ~/r/fish-shell>
```

Interestingly, the tests still pass.

**Comment by mqudsi** [comment]
WSL only.

**Comment by mqudsi** [comment]
git bisect traced it back to my own commit, be13ac353bfbd451f92b8037fa4fbf01cda73a95 😱.

Looking into it now.

**Comment by mqudsi** [comment]
That's weird, reverting that commit oddly does not resolve the problem.

**Comment by mqudsi** [comment]
Turns out some of the cleanup I removed to fix #4715, in particular, to remove needless calls to `setpgid` on a spawned child process because `posix_spawn` will set it itself exposed another WSL bug. Ugh.

https://github.com/Microsoft/WSL/issues/2997

**Comment by mqudsi** [comment]
So I just closed this with 0dd2607caccdae24b9bc7c94b8e8284d80b4e529.

I've never stopped thinking about this from when it came up. The results reported by @zanchey in #4715 were not what I had originally expected, and the fact that I've never been able to get this *not* to reproduce with glibc under WSL while it only rarely happens under Linux/glibc is annoying, but the workaround _should_ be ok. We only set the group if it's not yet set after `posix_spawn()`, though of course it *could* be set in between the time we check that and the time when our call to `setpgid()` actually goes through...

Ideally, we would dump `posix_spawn()` and just use `vfork()` directly to get guaranteed atomicity semantics here.
